### PR TITLE
Implement formvalidation repo

### PR DESCRIPTION
**Pull request by jordanjones243**
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

**Resolves:** #127, #129 

## Summary:

_Please summarize the scope of the changes you have submitted, what the intent of the work is and anything that describes the before/after state of the project._

* Implement `auro-formvalidation` repo
* Update JS docs
* Update dependencies
* Update styles to fix focus UI borders

## Type of change:

Please delete options that are not relevant.

- [x] New capability
- [x] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

**Comment by jordanjones243** [comment]
This PR will stay in draft mode until all form elements have `auro-formvalidation` implementation

File: src/auro-checkbox-group.js [hunk 1]
```diff
@@ -16,11 +16,15 @@
  * @attr {Boolean} noValidate - If set, disables auto-validation on blur.
  * @attr {Boolean} required - Populates the `required` attribute on the element. Used for client-side validation.
  * @attr {Boolean} horizontal - If set, checkboxes will be aligned horizontally.
+ * @event auroCheckboxGroup-validated - Notifies that the `validity` value has changed.
```
Comment (jordanjones243) [inline src/auro-checkbox-group.js:19]: Before release, figure out issue with doc build

File: src/auro-checkbox-group.js [hunk 2]
```diff
@@ -16,14 +24,13 @@
  * @attr {Boolean} noValidate - If set, disables auto-validation on blur.
  * @attr {Boolean} required - Populates the `required` attribute on the element. Used for client-side validation.
  * @attr {Boolean} horizontal - If set, checkboxes will be aligned horizontally.
+ * @slot legend - Allows for the legend to be overridden.
+ * @slot optionalLabel - Allows for the optional label to be overridden.
+ * @slot helpText - Allows for the helper text to be overridden.
+ * @fires auroCheckboxGroup-validated - Notifies that the `validity` value has changed.
+ * @fires auroCheckboxGroup-helpText - Notifies that the `setCustomValidity` value has changed.
```
Comment (blackfalcon) [inline src/auro-checkbox-group.js:31]:
```
checkbox/src/auro-checkbox-group.js
  30:1  warning  The type 'auroCheckboxGroup-validated' is undefined  jsdoc/no-undefined-types
  31:1  warning  The type 'auroCheckboxGroup-helpText' is undefined   jsdoc/no-undefined-types
```

Where are these methods defined?

File: src/auro-checkbox-group.js [hunk 3]
```diff
@@ -232,21 +203,39 @@ export class AuroCheckboxGroup extends LitElement {
...
-    }
+    this.validation.validate(this);
+  }
 
-    this.validate();
+  /**
+   * Helper method that handles attributes of checkboxes.
+   * @private
+   * @returns {void}
+   */
```
Comment (blackfalcon) [inline src/auro-checkbox-group.js:227]:
```suggestion
/**
 * Updates checkbox attributes for each item (private method).
 * Sets the 'error' attribute if there is an error, and adjusts 'disabled' and 'required' properties.
 *
 * @returns {void}
 * @private
 */
```

**Review by blackfalcon — CHANGES_REQUESTED** [review]

File: src/auro-checkbox-group.js (hunk 2 above)
Comment (jordanjones243) [reply src/auro-checkbox-group.js:31]: Those are events, not methods, but those events are dispatched from the `auro-formvalidation` repo.

File: demo/api.md [hunk 4]
```diff
@@ -14,17 +47,16 @@ Custom element for the purpose of allowing users to select one or more options o
 | [error](#error)    | `error`    | `Boolean` | false   | If set to true, sets an error state on the checkbox. |
 | [id](#id)       | `id`       | `String`  |         | Sets the individual `id` per element.            |
 | [name](#name)     | `name`     | `String`  |         | Accepts any string, `DOMString` representing the value of the input. |
-| [required](#required) | `required` | `Boolean` | false   | Populates the `required` attribute on the checkbox. Used for client-side validation. |
+| [required](#required) | `required` | `boolean` | false   |                                                  |
```
Comment (jason-capsule42) [inline demo/api.md:50]: Looks like the `required` prop still needs to get removed.

File: demo/api.md [hunk 5]
```diff
@@ -14,17 +47,16 @@ Custom element for the purpose of allowing users to select one or more options o
...
-|----------------|--------------------------------------|
-| [handleChange](#handleChange) | `(event: any): void`                 |
-| [handleInput](#handleInput)  | `(event: any): void`                 |
-| [invalid](#invalid)      | `(error: any): "true" \| "false"`    |
-| [isRequired](#isRequired)   | `(required: any): "true" \| "false"` |
+| Method         | Type                              |
+|----------------|-----------------------------------|
+| [handleChange](#handleChange) | `(event: any): void`              |
+| [handleInput](#handleInput)  | `(event: any): void`              |
+| [invalid](#invalid)      | `(error: any): "true" \| "false"` |
```
Comment (jason-capsule42) [inline demo/api.md:59]: Should this get removed as well since it's not the group that is validated?

File: docs/api.md [hunk 6]
```diff
@@ -11,17 +45,16 @@ Custom element for the purpose of allowing users to select one or more options o
 | `error`    | `error`    | `Boolean` | false   | If set to true, sets an error state on the checkbox. |
 | `id`       | `id`       | `String`  |         | Sets the individual `id` per element.            |
 | `name`     | `name`     | `String`  |         | Accepts any string, `DOMString` representing the value of the input. |
-| `required` | `required` | `Boolean` | false   | Populates the `required` attribute on the checkbox. Used for client-side validation. |
+| `required` | `required` | `boolean` | false   |                                                  |
```
Comment (jason-capsule42) [inline docs/api.md:48]: Missing JSDoc description

File: test/auro-checkbox.test.js [hunk 7]
```diff
@@ -225,11 +227,9 @@ describe('auro-checkbox', () => {
     expect(input.value).to.equal(expectedValue);
     expect(input.name).to.equal(expectedName);
     expect(input.type).to.equal('checkbox');
-    expect(input.getAttribute('aria-invalid')).to.equal('true');
-    expect(input.getAttribute('aria-required')).to.equal('true');
     expect(errorBorder).to.not.be.undefined;
     expect(el).dom.to.equal(`
-      <auro-checkbox id="${expectedId}" name="${expectedName}" value="${expectedValue}" error checked disabled required>
```
Comment (jason-capsule42) [inline test/auro-checkbox.test.js:232]: Is this being tested on the group now?

**Review by jason-capsule42 — CHANGES_REQUESTED** [review]

**Review by jason-capsule42 — COMMENTED** [review]
Also need to add in the new library utility that adds the element name as an attribute on both checkbox and group if it's been custom registered.

File: src/auro-checkbox-group.js [hunk 8]
```diff
@@ -255,20 +252,26 @@ export class AuroCheckboxGroup extends LitElement {
...
 
     if (changedProperties.has('required')) {
-      this.items.forEach((el) => {
-        el.required = this.required;
-      });
+      if (this.required) {
+        this.setAttribute('aria-required', true);
+      } else {
+        this.removeAttribute('aria-required');
+      }
```
Comment (jason-capsule42) [inline src/auro-checkbox-group.js:264]: I think this can and should be done in the constructor.

```js
/**
 * @private
 */
this.aria-required = this.required

aria-required: {
  type: Boolean,
  reflect: true
},
```

File: src/auro-checkbox-group.js [hunk 9]
```diff
@@ -255,20 +252,26 @@ export class AuroCheckboxGroup extends LitElement {
...
     }
 
     if (changedProperties.has('error')) {
-      this.validate();
+      if (this.error) {
+        this.setAttribute('aria-invalid', true);
+      } else {
+        this.removeAttribute('aria-invalid');
+      }
```
Comment (jason-capsule42) [inline src/auro-checkbox-group.js:272]: Same as for aria-required above

File: test/auro-checkbox.test.js [hunk 10]
```diff
@@ -15,6 +15,8 @@ describe('auro-checkbox-group', () => {
       <auro-checkbox-group
         horizontal
         required
+        aria-invalid
+        aria-required
```
Comment (jason-capsule42) [inline test/auro-checkbox.test.js:19]: Why are these added here? The test should be that the code adds them based on validityState and the presence of the `required` attribute.

File: test/auro-checkbox.test.js [hunk 11]
```diff
@@ -24,10 +26,12 @@ describe('auro-checkbox-group', () => {
 
     expect(el.horizontal).to.be.true;
     expect(el.required).to.be.true;
+    expect(el.ariaRequired).to.equal('true');
+    expect(el.ariaInvalid).to.equal('true');
```
Comment (jason-capsule42) [inline test/auro-checkbox.test.js:30]: We should test for `required` and `validity` as separate tests.

**Review by jason-capsule42 — CHANGES_REQUESTED** [review]

File: package.json [hunk 12]
```diff
@@ -114,8 +116,8 @@
             "./CHANGELOG.md",
             "./README.md",
             "./docs/api.md",
-            "./demo/demo.md",
-            "./demo/apiExamples.md"
+            "./demo/index.md",
+            "./demo/api.md"
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply package.json:120]: I don't think this should affect anything, does it? This is just changing the file names in the `package.json`, but the file names were already named this way in the repo.

File: apiExamples/error.html [hunk 13]
```diff
@@ -1 +0,0 @@
-<auro-checkbox value="error checkbox option" name="example3" id="checkbox-error1" error>Error checkbox option</auro-checkbox>
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply apiExamples/error.html:1]: We already have this example on the API page

File: src/auro-checkbox-group.js (hunk 9 above)
Comment (jordanjones243) [reply src/auro-checkbox-group.js:272]: Same as above.

File: docs/partials/api.md [hunk 14]
```diff
@@ -76,25 +76,6 @@ The `error` attribute used to set error state on the entire `<auro-checkbox-grou
 
 </auro-accordion>
 
-#### Checkbox with Error Outside of Group
-
-The `error` attribute used to set an error state on a single `<auro-checkbox>` element.
-
-**Note**: Error messages are not supported outside of an `<auro-checkbox-group>`.
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply docs/partials/api.md:83]: Added

**Review by jason-capsule42 — APPROVED** [review]
